src/handler.ts

```diff
@@ -1,33 +1,82 @@
 import { CW } from "./cloudwatch";
 import { Dynamo } from "./dynamodb";
 import { Category, CategoryConfiguration, CategoryServiceFactory, LogLevel } from "typescript-logging";
 import { ungzip } from "node-gzip";
-
+import AWSXRay from "aws-xray-sdk";
+import RE2 from "re2";
 
 CategoryServiceFactory.setDefaultConfiguration(new CategoryConfiguration(LogLevel.Info));
 export const handlerLogger = new Category("Handler");
 export const dynamoLogger = new Category("DynamoDB", handlerLogger);
 export const cwLogger = new Category("CloudWatch", handlerLogger);
 
+/**
+ * Decodes the base64 event data and decompresses it.
+ *
+ * @param {string} data Base64 encoded data from firehose
+ * @returns {CloudWatchLogsDecodedData} The decoded data
+ */
 async function decodeEventData(data: string): Promise<CloudWatchLogsDecodedData> {
-  const unzipped = String(await ungzip(Buffer.from(data, "base64")));
+  const unzipped = (await ungzip(Buffer.from(data, "base64"))).toString();
   handlerLogger.info(unzipped);
   return JSON.parse(unzipped);
 }
 
 /**
- * @param {CloudWatchLogsEvent} event
- * @param {Context} context
+ * The lambda handler for metrics collector, no transformation of the data happens as this only generates metrics.
+ *
+ * @param {FirehoseTransformationEvent} event The expected event from Firehose
+ * @param {Context | undefined} context The context of the running lambda
+ * @returns {FirehoseTransformationResult} The result to return to Firehose after processing.
  */
-export const handler: CloudWatchLogsHandler = async (event: CloudWatchLogsEvent, context: Context): Promise<void> => {
-  handlerLogger.info(`context: ${JSON.stringify(context)}`);
-  const log: CloudWatchLogsDecodedData = await decodeEventData(event.awslogs.data);
-  const cw = new CW();
-  if (/\/aws\/lambda\/activities-[\w-]+/.test(log.logGroup)) {
-    const dynamo = new Dynamo();
-    const [visits, oldVisits, openVisits] = await Promise.all([dynamo.getVisits(), dynamo.getOldVisits(), dynamo.getOpenVisits()]);
-    handlerLogger.info(await cw.sendVisits(visits, oldVisits, openVisits));
+export const handler = async (event: FirehoseTransformationEvent, context?: Context | undefined): Promise<FirehoseTransformationResult> => {
+  try {
+    const decodeSS = AWSXRay.getSegment()?.addNewSubsegment("decodeEvent");
+    handlerLogger.info(`context: ${JSON.stringify(context)}`);
+    const logs: CloudWatchLogsDecodedData[] = await Promise.all(
+      event.records.map((record) => {
+        return decodeEventData(record.data);
+      })
+    );
+    decodeSS?.addMetadata("decodedEvent", logs);
+    decodeSS?.close();
+    const cw = new CW();
+    const promises: Promise<void>[] = [];
+    const activityPattern = new RE2(/\/aws\/lambda\/activities-[\w-]+/);
+    const sendMetricsSS = AWSXRay.getSegment()?.addNewSubsegment("sendMetrics");
+    // Prevent activities metrics being sent more than once.
+    let activitiesSent = false;
```

Comment:
I think I will need a catch up on that one whenever you can do it please

Reply:
The code was written before would get batches of logs for a single microservice, now using firehose, there will likely be multiple microservice logs interspersed. There is no benefit to sending the same metrics milliseconds after the last.